### ISSUE TITLE
Postgres pod starts as secondary node instead of primary after Kubernetes cluster upgrade

### ISSUE THREAD

**Issue by nihaldivyam** [zalando/postgres-operator#2288]
Please provide the following information:

- **Which version of the Postgres Operator are you using?**
The version of the Postgres Operator being used is registry.opensource.zalan.do/acid/postgres-operator:v1.9.0.

- **Where is the Postgres Operator running? Is it on cloud or bare metal? Are you using Kubernetes or OpenShift?**
The Postgres Operator is running on bare metal.

- **Are you currently using the Postgres Operator in production?**
Yes, the Postgres Operator is currently being used in production.

- **What type of issue are you facing?**
This is a bug report.

Description:
We upgraded our Kubernetes cluster from version 1.24.8 to 1.26.3 by following the official docs. When we drained the node running the Postgres pod, it would respawn on another node without any issues. However, the Postgres pod started as a secondary node instead of the primary node, even though we had configured it to run in 1-master mode. As a result, the database rejected all connections from the application. Please see the error log below for more information:

```
2023-04-14 10:26:25,574 INFO: Lock owner: None; I am mattermost-pgsql-0                                                                                                                    │
│ 2023-04-14 10:26:25,574 INFO: starting as a secondary                                                                                                                                      │
│ 2023-04-14 10:26:25,664 INFO: postmaster pid=85678 
│ 2023-04-12 10:26:25 UTC [15597]: [8-1] 6436b473.3ced 0     HINT:  Future log output will appear in directory "../pg_log".                                                                  │
│ /var/run/postgresql:5432 - rejecting connections                                                                                                                                           │
│ /var/run/postgresql:5432 - rejecting connections                                                                                                                                           │
│ /var/run/postgresql:5432 - rejecting connections                                                                                                                                           │
│ /var/run/postgresql:5432 - rejecting connections                                                                                                                                           │
│ /var/run/postgresql:5432 - no response 

```

**Comment by FxKu** [comment]
So you're saying the respawned pod is not elected as the leader and remains the replica? Or does it happen automatically after some short time? Can you check what's in the Postgres logs?

**Comment by nihaldivyam** [comment]
@FxKu Yes the respawned pod is not elected as the leader and remains the replica, it happened when i drained the node where the Postgres was running, and when it respawned it ran as a replica, not a leader.

**Comment by yuvraj-vansure** [comment]
Hi @nihaldivyam, were you able to solve the issue? I'm also facing similar issue but with crunchydata/postgres-operator.

**Comment by nihaldivyam** [comment]
@FxKu, the problem persists. Despite restarting the PostgreSQL pod, it consistently runs as secondary even when there's only one pod. We attempted scaling it up to ensure at least one pod operates as primary, but this did not resolve the issue.

```
kubectl logs api-postgresql-0 -n random  --tail=15
  Date/time type storage: 64-bit integers
  Float8 argument passing: by value
  Data page checksum version: 0
  Mock authentication nonce: 1eae14c3d9ef8fdbdcde04620b17425bf9b9434efe03cc3cd1c1ecf4d0fa9eb3

2023-12-27 13:09:17,926 INFO: Lock owner: None; I am api-postgresql-0
2023-12-27 13:09:17,927 INFO: starting as a secondary
2023-12-27 13:09:18,033 INFO: postmaster pid=167017
/var/run/postgresql:5432 - no response
2023-12-27 13:09:18 UTC [167017]: [1-1] 658c21fe.28c69 0     LOG:  Auto detecting pg_stat_kcache.linux_hz parameter...
2023-12-27 13:09:18 UTC [167017]: [2-1] 658c21fe.28c69 0     LOG:  pg_stat_kcache.linux_hz is set to 333333
2023-12-27 13:09:18 UTC [167017]: [3-1] 658c21fe.28c69 0     LOG:  redirecting log output to logging collector process
2023-12-27 13:09:18 UTC [167017]: [4-1] 658c21fe.28c69 0     HINT:  Future log output will appear in directory "../pg_log".
/var/run/postgresql:5432 - rejecting connections
/var/run/postgresql:5432 - rejecting connections

```

**Comment by bumarcell** [comment]
We're having the same problem. No idea why the **only pod** is starting as secondary, knowing there's no lock owner 😖 

`2024-04-26 09:33:54,713 INFO: Lock owner: None; I am zeus-postgres-0
2024-04-26 09:33:54,714 INFO: starting as a secondary
2024-04-26 09:33:54,813 INFO: postmaster pid=26709
/var/run/postgresql:5432 - no response
2024-04-26 09:33:54 UTC [26709]: [1-1] 662b7502.6855 0     LOG:  Auto detecting pg_stat_kcache.linux_hz parameter...
2024-04-26 09:33:54 UTC [26709]: [2-1] 662b7502.6855 0     LOG:  pg_stat_kcache.linux_hz is set to 1000000
2024-04-26 09:33:54 UTC [26709]: [3-1] 662b7502.6855 0     LOG:  redirecting log output to logging collector process
2024-04-26 09:33:54 UTC [26709]: [4-1] 662b7502.6855 0     HINT:  Future log output will appear in directory "../pg_log".
/var/run/postgresql:5432 - rejecting connections
/var/run/postgresql:5432 - rejecting connections
/var/run/postgresql:5432 - no response`

**Comment by bumarcell** [comment]
Update: the problem seems to have been caused by old data in the s3 bucket. After having removed those the master was promoted.

**Comment by Ali-Rezaalii** [comment]
Thank you so much,The issue has been resolved successfully, and everything is now working as expected

**Comment by Ali-Rezaalii** [comment]
> Update: the problem seems to have been caused by old data in the s3 bucket. After having removed those the master was promoted.

Thank you so much,The issue has been resolved successfully, and everything is now working as expected

**Comment by nihaldivyam** [comment]
@bumarcell @Ali-Rezaalii What exact data you guys removed the wal folder ?

**Comment by Ali-Rezaalii** [comment]
> @bumarcell @Ali-Rezaalii What exact data you guys removed the wal folder ? remove pg_wal old file with pg_controldata & pg_archivecleanup

@nihaldivyam — here is the exact data we removed from `pg_wal` and how we identified it:

**1. Find the checkpoint REDO WAL file (the oldest file you must keep):**
```bash
pg_controldata $PGDATA | egrep "Latest checkpoint location|Latest checkpoint's REDO WAL file"
# e.g.  Latest checkpoint's REDO WAL file: 000000010000000000000005
```

**2. Dry-run the cleanup first** (`-d` only prints what would be removed):
```bash
pg_archivecleanup -d $PGDATA/pg_wal 000000010000000000000005
```

**3. Remove the old segments (everything before the REDO WAL file):**
```bash
pg_archivecleanup $PGDATA/pg_wal 000000010000000000000005
```

`pg_archivecleanup` deletes every WAL segment in the target directory that is *older than* the file you pass as the second argument — so after it runs, `pg_wal` contains only the REDO segment and everything newer. That is the "exact data" we removed: obsolete completed/`.partial` segments from before the latest checkpoint, plus after a k8s upgrade failover you may also need to drop the *old timeline's* segments (`00000002…` vs `00000001…` — keep only the current timeline found in `pg_controldata`).

**Also clean these leftovers, they are what keeps the pod starting as a secondary:**
```bash
rm -f $PGDATA/standby.signal $PGDATA/pg_wal/recovery.signal
```

**Procedure we followed (do this with Postgres stopped):**
1. `patronictl stop` (or scale the container) so no postmaster is writing.
2. `tar czf /backup/pg_wal-backup-$(date +%F).tar.gz $PGDATA/pg_wal` — always keep a copy before deleting.
3. Run steps 1–3 above against the live `$PGDATA`.
4. Confirm: `pg_controldata $PGDATA | grep "Database cluster state"` → should be consistent; restart the pod and Patroni promotes it as primary in the configured 1-master cluster.
5. If a stale replica still refuses to rejoin, `patronictl reinit <cluster> <replica>` is the clean path instead of manual WAL surgery.

Caution: never delete the segment named by `Latest checkpoint's REDO WAL file` or anything newer, and verify the restored timeline with `SELECT pg_last_wal_replay_lsn();` on replicas afterwards.
